### PR TITLE
bpo-44165: optimise sqlite3 statement preparation by passing string size

### PR DESCRIPTION
**Pull request by erlend-aasland**
Quoting the SQLite docs, https://sqlite.org/c3ref/prepare.html:
"If the caller knows that the supplied string is nul-terminated, then
there is a small performance advantage to passing an nByte parameter
that is the number of bytes in the input string including the
nul-terminator."

https://bugs.python.org/issue44165

### CODE REVIEW

**Comment by erlend-aasland** [comment]
@pablogsal I recon this can pass as _skip news_. This is an internal micro-optimisation only. No user visible change.

File: Modules/_sqlite/statement.c [hunk 1]
```diff
@@ -96,7 +96,7 @@ int pysqlite_statement_create(pysqlite_Statement* self, pysqlite_Connection* con
     Py_BEGIN_ALLOW_THREADS
     rc = sqlite3_prepare_v2(connection->db,
                             sql_cstr,
-                            -1,
+                            sql_cstr_len + 1,
```
Comment (pablogsal) [inline Modules/_sqlite/statement.c]: @erlend-aasland Can you check out the compiler warnings?

File: Modules/_sqlite/cursor.c [hunk 2]
```diff
@@ -706,13 +707,16 @@ pysqlite_cursor_executescript(pysqlite_Cursor *self, PyObject *script_obj)
...
 
+    Py_ssize_t size_incl_null = script_length + 1;
     while (1) {
+        const char *tail;
+
         Py_BEGIN_ALLOW_THREADS
         rc = sqlite3_prepare_v2(self->connection->db,
                                 script_cstr,
-                                -1,
+                                size_incl_null,
```
Comment (pablogsal) [inline Modules/_sqlite/cursor.c]: Same here

File: Modules/_sqlite/statement.c (hunk 1 above)
Comment (erlend-aasland) [reply Modules/_sqlite/statement.c]: Oops, thanks for the heads up!

**Review by pablogsal — APPROVED** [review]

File: Modules/_sqlite/cursor.c [hunk 3]
```diff
@@ -690,10 +691,17 @@ pysqlite_cursor_executescript(pysqlite_Cursor *self, PyObject *script_obj)
...
-        script_cstr = PyUnicode_AsUTF8(script_obj);
+        script_cstr = PyUnicode_AsUTF8AndSize(script_obj, &sql_len);
         if (!script_cstr) {
             return NULL;
         }
+
+        int max_length = sqlite3_limit(self->connection->db,
+                                       SQLITE_LIMIT_LENGTH, -1);
+        if (sql_len >= max_length) {
+            PyErr_SetString(pysqlite_DataError, "query string is too large");
```
Comment (pablogsal) [inline Modules/_sqlite/cursor.c:702]: @erlend-aasland Can you make another PR with some new test to exercise these errors?

File: Modules/_sqlite/statement.c [hunk 4]
```diff
@@ -66,6 +66,12 @@ int pysqlite_statement_create(pysqlite_Statement* self, pysqlite_Connection* con
         rc = PYSQLITE_SQL_WRONG_TYPE;
         return rc;
     }
+
+    int max_length = sqlite3_limit(connection->db, SQLITE_LIMIT_LENGTH, -1);
+    if (sql_cstr_len >= max_length) {
+        PyErr_SetString(pysqlite_DataError, "query string is too large");
```
Comment (pablogsal) [inline Modules/_sqlite/statement.c:72]: Same as https://github.com/python/cpython/pull/26206/files#r643910263

File: Modules/_sqlite/cursor.c (hunk 3 above)
Comment (erlend-aasland) [reply Modules/_sqlite/cursor.c:702]: Of course.

**Comment by erlend-aasland** [comment]
This PR was broken by fffa0f92adaaed0bcb3907d982506f78925e9052. I'll resolve the issues in a new PR.
